### ISSUE TITLE
The speed of FirstLevelModel fit() does not depend on mask size and compute time scales nonlinearly with image volume

### ISSUE THREAD

**Issue by themeo** [nilearn/nilearn#3592]
This is not strictly a bug but I believe Nilearn could more efficiently compute GLMs by restricting itself to the provided mask. 
I use Nilearn version: 0.10.0

### Specific example:
I have two functional images of the same brain, outputs from `fmriprep`, one in T1 and one in functional space.
The images have the same resolution and data precision but differ in the overall image size `(104, 104, 66, 390)` vs. `(68, 85, 69, 390)` because T1 is cropped to the brain.

Running a simple `FirstLevelModel` on the two images, with a V1 mask (~4000 voxels in both), leads to a 10-fold difference in compute times: 3s (smaller, T1) vs. 30s (larger, functional). All extra actions (scaling, standardization, filtering, etc) are turned off. 

This seems to show that:
* computations are not restricted to the mask but are applied to the whole image 
* compute time does not linearly scale with brain volume (volume of the larger image is less than 2x of the smaller image, compute time difference is 10x)

### Expected behavior
Until now I assumed that `FirstLevelModel` first extracts masked data using e.g. `NiftiMasker` and only then computes the GLM voxelwise. 

Perhaps this is intended behaviour for other reasons, but then perhaps it could be mentioned in the documentation. 

### Example of a temporary workaround

Unfortunately `FirstLevelModel` cannot operate on `NiftiMasker` objects, so my temporary workaround uses image cropping:

```python
mask_crop, slicer = nilearn.image.crop_img(mask_file, return_offset=True)
fmri_glm = FirstLevelModel(..., mask_img=mask_crop, ...)
img = nib.load(imgfile)
img = nilearn.image.new_img_like(mask_crop, np.asarray(img.dataobj)[slicer])
```

In my use case (LSS-type regression, ~300 events) this speeds up computations from 50m to 1m.

**Comment by bthirion** [comment]
Thx for reporting !
In both cases, do you provide the mask that specifies  the region of interest ? Normally, since the computations are done on masked data, there should be little difference when the mask is provided.

**Comment by themeo** [comment]
Yes, in both cases I provide a mask. The masks differ slightly because they come from different spaces (T1 vs BOLD), but they have a very similar number of voxels (~4000).

**Comment by bthirion** [comment]
The next step would be to reproduce this on a dataset provided by Nilearn, e.g. by cropping an image on which a GLM is run. Would you mind doing this ? That would help a lot. Let us know if you don't want or cannot do it.
Best,

**Comment by themeo** [comment]
Sure thing. I looped the GLM because with such a small image (my images are 5x larger) it computed very fast even in the uncropped version. 

```python
from nilearn import datasets
from nilearn.glm.first_level import FirstLevelModel
import nilearn.image
import nilearn.datasets
import numpy as np
import pandas as pd
import time


dataset = datasets.fetch_development_fmri(n_subjects=1)
img = nilearn.image.load_img(dataset.func)

# Create a fake cubic mask
mask = np.zeros(img.shape[:-1])
mask[20:30, 20:30, 20:30] = 1
mask = nilearn.image.new_img_like(img, mask)

# Create fake events
events = pd.DataFrame()
data = []
for ix, onset in enumerate(np.arange(3, img.shape[-1], 5)):
    data.append([onset, 0.5, str(ix)])
events = pd.DataFrame(data, columns=['onset', 'duration', 'trial_type'])


# GLM with uncropped mask
glm1 = FirstLevelModel(t_r=1., drift_model=None, noise_model='ols', mask_img=mask)
img.get_fdata() # Make sure it is cached
t1 = time.process_time()
for ix in range(100):
    glm1.fit(img, events)
print (time.process_time() - t1)
# runtime: 59.53125


# GLM with cropped mask
mask_crop, slicer = nilearn.image.crop_img(mask, return_offset=True)
glm2 = FirstLevelModel(t_r=1., drift_model=None, noise_model='ols', mask_img=mask_crop)
img_crop = nilearn.image.new_img_like(mask_crop, np.asarray(img.dataobj)[slicer])
img_crop.get_fdata()
t1 = time.process_time()
for ix in range(100):
    glm2.fit(img_crop, events)
print (time.process_time() - t1) 
# runtime: 30.1875
```
